src/modules/musig/session_impl.h

```diff
@@ -0,0 +1,632 @@
+}
+
+int secp256k1_musig_aggnonce_serialize(const secp256k1_context* ctx, unsigned char *out66, const secp256k1_musig_aggnonce* nonce) {
+    return secp256k1_musig_pubnonce_serialize(ctx, out66, (secp256k1_musig_pubnonce*) nonce);
+}
+
+int secp256k1_musig_aggnonce_parse(const secp256k1_context* ctx, secp256k1_musig_aggnonce* nonce, const unsigned char *in66) {
+    return secp256k1_musig_pubnonce_parse(ctx, (secp256k1_musig_pubnonce*) nonce, in66);
+}
+
+int secp256k1_musig_partial_sig_serialize(const secp256k1_context* ctx, unsigned char *out32, const secp256k1_musig_partial_sig* sig) {
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(out32 != NULL);
+    ARG_CHECK(sig != NULL);
+    memcpy(out32, &sig->data[4], 32);
+    return 1;
+}
+
+int secp256k1_musig_partial_sig_parse(const secp256k1_context* ctx, secp256k1_musig_partial_sig* sig, const unsigned char *in32) {
+    secp256k1_scalar tmp;
+    int overflow;
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(sig != NULL);
+    ARG_CHECK(in32 != NULL);
+
+    secp256k1_scalar_set_b32(&tmp, in32, &overflow);
+    if (overflow) {
+        secp256k1_scalar_clear(&tmp);
+        return 0;
+    }
+    secp256k1_musig_partial_sig_save(sig, &tmp);
+    secp256k1_scalar_clear(&tmp);
+    return 1;
+}
+
+/* Normalizes the x-coordinate of the given group element. */
+static int secp256k1_xonly_ge_serialize(unsigned char *output32, secp256k1_ge *ge) {
+    if (secp256k1_ge_is_infinity(ge)) {
+        return 0;
+    }
+    secp256k1_fe_normalize_var(&ge->x);
+    secp256k1_fe_get_b32(output32, &ge->x);
+    return 1;
+}
+
+static void secp256k1_nonce_function_musig(secp256k1_scalar *k, const unsigned char *session_id, const unsigned char *key32, const unsigned char *msg32, const unsigned char *agg_pk32, const unsigned char *extra_input32) {
+    secp256k1_sha256 sha;
+    unsigned char seed[32];
+    unsigned char i;
+    enum { n_extra_in = 4 };
+    const unsigned char *extra_in[n_extra_in];
+
+    /* TODO: this doesn't have the same sidechannel resistance as the BIP340
+     * nonce function because the seckey feeds directly into SHA. */
+    secp256k1_sha256_initialize_tagged(&sha, (unsigned char*)"MuSig/nonce", sizeof("MuSig/nonce"));
+    secp256k1_sha256_write(&sha, session_id, 32);
+    extra_in[0] = key32;
+    extra_in[1] = agg_pk32;
+    extra_in[2] = msg32;
+    extra_in[3] = extra_input32;
+    for (i = 0; i < n_extra_in; i++) {
+        unsigned char marker;
+        if (extra_in[i] != NULL) {
+            marker = 1;
+            secp256k1_sha256_write(&sha, &marker, 1);
+            secp256k1_sha256_write(&sha, extra_in[i], 32);
+        } else {
+            marker = 0;
+            secp256k1_sha256_write(&sha, &marker, 1);
+        }
+    }
+    secp256k1_sha256_finalize(&sha, seed);
+
+    for (i = 0; i < 2; i++) {
+        unsigned char buf[32];
+        secp256k1_sha256_initialize(&sha);
+        secp256k1_sha256_write(&sha, seed, 32);
+        secp256k1_sha256_write(&sha, &i, 1);
```

Comment:
```suggestion
        secp256k1_sha256_write(&sha, &i, sizeof(i));
```

The current code is correct but it crucially depends on `i` being `unsigned char`. When reading this, I first implicitly assumed that `i` is `int` (because that's normal for loop counters) and thought that we have a very subtle bug here on bigendian system where then the byte at address `&i` is always 0x00... I fear that someone will change the type of `i` without noticing and then we'll have this bug. 

The above suggestion makes the code more robust against this but it may be a good idea to additionally use an `int` as loop counter and convert it to `unsigned char` explicitly, or alternatively make `i` an `int` but use a new distinct loop variable for this loop here.

Reply:
That's a good point. Accepted your suggestion. Not sure about doing more than that as it seems to be getting quite complex.